src/app/modules/item/services/threads.service.ts

```diff
@@ -44,37 +98,115 @@ interface TokenData {
   canWatchParticipant: boolean,
 }
 
-type Action = 'open-thread' | 'close-thread' | 'follow';
+type ThreadAction =
+  | { action: 'open-thread', history: ActivityLog[] }
+  | { action: 'close-thread' }
+  | { action: 'follow' }
+  | { action: 'unfollow' }
+  | { action: 'thread-status' }
+  | { action: 'send-message', message: string };
 
 @Injectable({
   providedIn: 'root',
 })
-export class ThreadService {
+export class ThreadService implements OnDestroy {
+
+  events$: Observable<ThreadEventMessage[]>;
+  status$: Observable<ThreadStatusMessage['status']>;
 
+  private newEvents$: Observable<ThreadEventMessage[]>;
+  private clearEvents$ = new Subject<void>();
   private tokenData?: TokenData;
   private socket: WebSocketSubject<unknown>;
-  messages$: Observable<ThreadEvent[]>;
+  private incomingMessages$: Observable<ThreadMessage[]>;
+  private subscriptions: Subscription[] = [];
 
-  constructor() {
+  constructor(
+    private activityLogService: ActivityLogService,
+  ) {
     if (!appConfig.forumServerUrl) throw new Error('cannot instantiate threads service when no forum server url');
     this.socket = webSocket(appConfig.forumServerUrl);
-    this.messages$ = this.socket.pipe(
-      decodeSnakeCase(D.array(threadEventDecoder)),
-      scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]),
+
+    this.incomingMessages$ = this.socket.pipe(
+      decodeSnakeCase(D.array(threadMessageDecoder)),
       catchError(() => EMPTY), // ignore undecoded messages
     );
+
+    this.newEvents$ = this.incomingMessages$.pipe(map(messages => messages.filter(isThreadEventMessage)));
+
+    this.events$ = this.clearEvents$.pipe(
+      startWith(undefined),
+      switchMap(() => this.newEvents$.pipe(scan((oldEvents, newEvents) => [ ...oldEvents, ...newEvents ]))),
+      map(messages => messages.sort((a, b) => a.time.valueOf() - b.time.valueOf())), // sort by date ascending
```

Comment:
If 2 messages have the same timestamp, it means they are the same, and that should not happen.

Reply:
In your code until L3 it is "events" not "message".  (is the naming change on purpose btw?) So a "follow" and "open thread" at the same time would not "means they are the same".
So I can imagine there may be 2 events occurring at the end moment.. just because you do not control what you get from activities... for instance having a score change event and a validation event at the same time.

Reply:
> in your code until L3 it is "events" not "message". (is the naming change on purpose btw?)

The naming is on purpose and will change since the `thread-status` action will become an http endpoint and won't be sent as a message anymore. When this work will be done, the frontend will only deal with events, which are in this context thread events. (Side note: although "unfollow" is not stored, it remains a thread event).

> So a "follow" and "open thread" at the same time would not "means they are the same".

This cannot happen due to dynamodb storage: 2 thread events - that are stored - cannot have the same pk+time couple since this couple is unique in dynamodb. If we were to add 2 entries with the same pk+time, the last entry would override the other. This issue is handled on the backend except for the thread events imported from history. Can you check in the DB if there are a lot of logs having the same itemId+userId+datetime values? (datetime precision is ms).

Reply:
For the message/events thing: 
- is there a clear definition of what is an event and a message (and if one includes the other one) somewhere is file? I remember you explain me what I cannot remember the difference anymore.
- so the code above, we work with events on line 139, then work with "messages" on line 140 and then store the results in "event$"... so that's is correct?

> This issue is handled on the backend except for the thread events imported from history. Can you check in the DB if there are a lot of logs having the same itemId+userId+datetime values? (datetime precision is ms).

There is no such thing as logs in the DB, it is computed logs. But nothing prevent them to have a same time so it will happen, so it has to be handled.

Reply:
> For the message/events thing

I changed the thread-status to an http service instead of socket, so from now on the frontend only deals with events, no more "message/events" distinction to make. Every socket message is a thread event.

---

For the activity logs ok, I'll add milliseconds to duplicates on the backend side.

Reply:
> For the activity logs ok, I'll add milliseconds to duplicates on the backend side.

Fixed on the backend, you can pull master